ironfish/src/mining/manager.ts

```diff
@@ -53,6 +66,130 @@ export class MiningManager {
+    if (!this.templateStream) {
+      const onNewBlockTemplate = new Event<[SerializedBlockTemplate]>()
+      onNewBlockTemplate.on(listener)
+      this.templateStream = { onNewBlockTemplate }
+
+      // Send an initial block template to the requester so they can begin working immediately
+      void this.chain.getBlock(this.chain.head).then((currentBlock) => {
+        if (currentBlock) {
+          this.onConnectedBlock(currentBlock)
+        }
+      })
+
+      return
+    }
+
+    if (this.templateStream.mostRecent) {
+      listener(this.templateStream.mostRecent.template)
+    }
+
+    this.templateStream.onNewBlockTemplate.on(listener)
+  }
+
+  offNewBlockTemplate(listener: (template: SerializedBlockTemplate) => void): void {
+    if (this.templateStream) {
+      this.templateStream.onNewBlockTemplate.off(listener)
+      if (this.templateStream.onNewBlockTemplate.isEmpty) {
+        this.templateStream = undefined
+      }
+    }
+  }
+
+  streamBlockTemplate(
+    currentBlock: Block,
+    template: SerializedBlockTemplate,
+    priority: number,
+  ): void {
+    // If there are not listeners for new blocks, return early
+    if (!this.templateStream) {
+      return
+    }
+
+    // If the new template is not building on top of a heavier block, return early
+    if (
+      this.templateStream.mostRecent &&
+      isBlockHeavier(this.templateStream.mostRecent.prevBlock.header, currentBlock.header)
+    ) {
+      return
+    }
+
+    // If new template was created after the current template, return early
+    if (this.templateStream.mostRecent && priority < this.templateStream.mostRecent.priority) {
+      return
+    }
+
+    this.templateStream.onNewBlockTemplate.emit(template)
+    this.templateStream.mostRecent = {
+      template,
+      priority,
+      prevBlock: currentBlock,
+    }
+  }
+
+  private onConnectedBlock(currentBlock: Block): void {
+    const connectedAt = BenchUtils.start()
+
+    // If we mine when we're not synced, then we will mine a fork no one cares about
+    if (!this.node.chain.synced && !this.node.config.get('miningForce')) {
+      return
+    }
+
+    // If we mine when we're not connected to anyone, then no one will get our blocks
+    if (!this.node.peerNetwork.isReady && !this.node.config.get('miningForce')) {
+      return
+    }
+
+    // If there are not listeners for new blocks, then return early
+    if (!this.templateStream) {
+      return
+    }
+
+    // TODO: Where should this check on account be? Do we want to re-query the default account on every connected block?
+    // Or do we want to set the account on node startup?
+    const account = this.node.wallet.getDefaultAccount()
+    if (!account) {
+      this.node.logger.info('Cannot mine without an account')
+      return
+    }
+
+    if (account.spendingKey === null) {
+      this.node.logger.info('Account must have spending key in order to mine')
+      return
+    }
+    AssertSpending(account)
```

Comment:
We do but `account.spendingKey === null` for some reason doesn't cut it for the typescript compiler. AssertSpending(account) throws an error which we don't want on a chain.onConnect() event or it could crash. So I did both. Let me see if I can find a way to make this cleaner though

Reply:
Added a helper function to make this more clear